### ISSUE TITLE
s2idle leads to an array-index-out-of-bounds in dswexec.c

### ISSUE THREAD

**Issue by OliverUrbann** [open-acpica/acpica#725]
Hi all,

Going into s2idle on a Surface Book 1 or Surface Pro 4 leads to the following stack trace. In my case it's an Ubuntu 20.04 LTS.

```
================================================================================
[  294.673738] UBSAN: array-index-out-of-bounds in drivers/acpi/acpica/dswexec.c:401:12
[  294.673748] index -1 is out of range for type 'acpi_operand_object *[9]'
[  294.673755] CPU: 3 PID: 6477 Comm: systemd-sleep Tainted: G         C        5.14.9-surface-ubsan-test #1
[  294.673762] Hardware name: Microsoft Corporation Surface Book/Surface Book, BIOS 92.3748.768 05.04.2021
[  294.673765] Call Trace:
[  294.673771]  dump_stack_lvl+0x4a/0x5f
[  294.673784]  dump_stack+0x10/0x12
[  294.673792]  ubsan_epilogue+0x9/0x50
[  294.673798]  __ubsan_handle_out_of_bounds+0x6f/0x80
[  294.673805]  acpi_ds_exec_end_op+0x1a0/0x79a
[  294.673812]  acpi_ps_parse_loop+0x7f5/0x8cc
[  294.673820]  acpi_ps_parse_aml+0x1bb/0x55d
[  294.673828]  acpi_ps_execute_method+0x20f/0x2d1
[  294.673836]  acpi_ns_evaluate+0x34d/0x4ef
[  294.673841]  acpi_evaluate_object+0x210/0x3da
[  294.673848]  acpi_evaluate_dsm+0xaa/0x120
[  294.673857]  ? flush_workqueue+0x19b/0x3e0
[  294.673864]  acpi_sleep_run_lps0_dsm+0x5a/0xc0
[  294.673873]  acpi_s2idle_restore_early+0x62/0x110
[  294.673881]  ? acpi_s2idle_restore_early+0x62/0x110
[  294.673887]  suspend_devices_and_enter+0x2a1/0x800
[  294.673895]  pm_suspend+0x2e5/0x420
[  294.673900]  state_store+0x85/0xf0
[  294.673905]  kobj_attr_store+0x12/0x20
[  294.673913]  sysfs_kf_write+0x3c/0x50
[  294.673921]  kernfs_fop_write_iter+0x13c/0x1b0
[  294.673927]  new_sync_write+0x117/0x1b0
[  294.673937]  vfs_write+0x1ea/0x250
[  294.673945]  ksys_write+0xa7/0xe0
[  294.673953]  __x64_sys_write+0x1a/0x20
[  294.673961]  do_syscall_64+0x5b/0xb0
[  294.673967]  ? syscall_exit_to_user_mode+0x2a/0x40
[  294.673974]  ? do_syscall_64+0x67/0xb0
[  294.673979]  ? do_syscall_64+0x67/0xb0
[  294.673983]  ? asm_exc_page_fault+0x8/0x30
[  294.673992]  entry_SYSCALL_64_after_hwframe+0x44/0xae
[  294.674000] RIP: 0033:0x7fdd5072c1e7
[  294.674007] Code: 64 89 02 48 c7 c0 ff ff ff ff eb bb 0f 1f 80 00 00 00 00 f3 0f 1e fa 64 8b 04 25 18 00 00 00 85 c0 75 10 b8 01 00 00 00 0f 05 <48> 3d 00 f0 ff ff 77 51 c3 48 83 ec 28 48 89 54 24 18 48 89 74 24
[  294.674012] RSP: 002b:00007fffdcfda2b8 EFLAGS: 00000246 ORIG_RAX: 0000000000000001
[  294.674019] RAX: ffffffffffffffda RBX: 0000000000000004 RCX: 00007fdd5072c1e7
[  294.674023] RDX: 0000000000000004 RSI: 00007fffdcfda370 RDI: 0000000000000004
[  294.674026] RBP: 00007fffdcfda370 R08: 0000000000000004 R09: 000000000000000d
[  294.674029] R10: 0000560dbe6e1128 R11: 0000000000000246 R12: 0000000000000004
[  294.674032] R13: 0000560dc03a72d0 R14: 0000000000000004 R15: 00007fdd508078a0
[  294.674038] ================================================================================
```

**Comment by acpibob** [comment]
Ok, I’ll try to take a look at it tomorrow.

**Comment by OliverUrbann** [comment]
I'm also not an expert for Linux kernel debugging. Maybe @qzed knows how to create a useful trace?

**Comment by acpibob** [comment]
Using the debug_level and debug_layer is the brute force way of getting a trace – since it will capture everything that the ACPICA subsystem is doing.  I believe that there is a way to trap certain control methods and trace their execution – but I don’t remember what it is on Linux. Perhaps the documentation specifies how to do this.  Anyway, it is apparently the _DSM method that we want to trace – all instances of it.

**Comment by daniel-ziegler** [comment]
I'm seeing the same warning on a ROG Zephyrus M16.

**Comment by acpibob** [comment]
I'm having trouble reproducing this problem. Can you give me the version of ACPICA you are using (should appear in the dmesg.log)?

**Comment by daniel-ziegler** [comment]
Here's what I have in dmesg; not sure if the version is there.
```
[ 3685.829337] PM: suspend entry (s2idle)
[ 3685.833713] Filesystems sync: 0.004 seconds
[ 3687.344503] rfkill: input handler enabled
[ 3687.550931] Freezing user space processes ... (elapsed 0.002 seconds) done.
[ 3687.553475] OOM killer disabled.
[ 3687.553476] Freezing remaining freezable tasks ... (elapsed 0.001 seconds) done.
[ 3687.554703] printk: Suspending console(s) (use no_console_suspend to debug)
[ 3688.276303] ACPI: EC: interrupt blocked
[ 3688.359129] ================================================================================
[ 3688.359131] UBSAN: array-index-out-of-bounds in /home/kernel/COD/linux/drivers/acpi/acpica/dswexec.c:401:12
[ 3688.359135] index -1 is out of range for type 'acpi_operand_object *[9]'
[ 3688.359138] CPU: 14 PID: 714 Comm: kworker/u32:4 Tainted: P           OE     5.15.13-051513-generic #202201050731
[ 3688.359141] Hardware name: ASUSTeK COMPUTER INC. ROG Zephyrus M16 GU603HM_GU603HM/GU603HM, BIOS GU603HM.315 12/09/2021
[ 3688.359143] Workqueue: events_unbound async_run_entry_fn
[ 3688.359151] Call Trace:
[ 3688.359153]  <TASK>
[ 3688.359157]  show_stack+0x52/0x58
[ 3688.359164]  dump_stack_lvl+0x4a/0x5f
[ 3688.359171]  dump_stack+0x10/0x12
[ 3688.359174]  ubsan_epilogue+0x9/0x45
[ 3688.359178]  __ubsan_handle_out_of_bounds.cold+0x44/0x49
[ 3688.359182]  acpi_ds_exec_end_op+0x1a1/0x798
[ 3688.359187]  acpi_ps_parse_loop+0x587/0x660
[ 3688.359191]  acpi_ps_parse_aml+0x1af/0x552
[ 3688.359194]  acpi_ps_execute_method+0x208/0x2ca
[ 3688.359196]  acpi_ns_evaluate+0x34e/0x4f0
[ 3688.359201]  acpi_evaluate_object+0x18e/0x3b4
[ 3688.359204]  __acpi_power_off+0x26/0x70
[ 3688.359209]  acpi_power_off_unlocked+0x50/0x90
[ 3688.359212]  acpi_power_transition+0x102/0x1a0
[ 3688.359215]  ? acpi_ns_get_node+0xaa/0xb8
[ 3688.359218]  acpi_device_set_power+0x2cb/0x360
[ 3688.359222]  acpi_pci_set_power_state+0x98/0x130
[ 3688.359228]  pci_platform_power_transition+0x6e/0xa0
[ 3688.359234]  pci_set_power_state+0x65/0x1a0
[ 3688.359236]  ? pci_disable_ptm+0x7f/0x90
[ 3688.359240]  pci_prepare_to_sleep+0x6f/0xd0
[ 3688.359243]  ? pci_pm_suspend_late+0x40/0x40
[ 3688.359245]  pci_pm_suspend_noirq+0xda/0x2e0
[ 3688.359247]  ? pci_pm_suspend_late+0x40/0x40
[ 3688.359249]  dpm_run_callback+0x69/0x120
[ 3688.359253]  __device_suspend_noirq+0x8f/0x280
[ 3688.359255]  async_suspend_noirq+0x23/0x60
[ 3688.359257]  async_run_entry_fn+0x30/0x120
[ 3688.359259]  process_one_work+0x228/0x3d0
[ 3688.359264]  worker_thread+0x53/0x410
[ 3688.359268]  kthread+0x11b/0x140
[ 3688.359271]  ? process_one_work+0x3d0/0x3d0
[ 3688.359274]  ? set_kthread_struct+0x50/0x50
[ 3688.359277]  ret_from_fork+0x1f/0x30
[ 3688.359283]  </TASK>
[ 3688.359284] ================================================================================
[ 4089.581769] ACPI: EC: interrupt unblocked
[ 4090.249975] pcieport 10000:e0:06.0: can't derive routing for PCI INT A
[ 4090.249978] nvme 10000:e1:00.0: PCI INT A: no GSI
[ 4090.323763] nvme nvme0: 16/0/0 default/read/poll queues
[ 4090.683779] OOM killer enabled.
[ 4090.683779] Restarting tasks ... 
[ 4090.683938] mei_hdcp 0000:00:16.0-b638ab7e-94e2-4ea2-a552-d1c54b627f04: bound 0000:00:02.0 (ops i915_hdcp_component_ops [i915])
[ 4090.691401] done.
[ 4090.697874] thermal thermal_zone7: failed to read out thermal zone (-61)
[ 4090.701809] PM: suspend exit
```

**Comment by satmandu** [comment]
Here are the ACPI messages in dmesg from 5.17-rc5:
sudo dmesg | grep ACPI > dmesg-ACPI.txt
[dmesg-ACPI.txt](https://github.com/acpica/acpica/files/8147030/dmesg-ACPI.txt)

**Comment by carter-thaxton** [comment]
Seeing the same thing on shutdown of a NUC10.  Reproducible at every shutdown.  Nothing new in the dmesg log than what's already been reported here.

**Comment by leodotcloud** [comment]
This issue happens consistently on my laptop when I try to wake up by opening the lid/screen after it had been sleeping.

Hardware: MSI GS66 
OS: Ubuntu 22.04.

Appreciate it if anyone can provide pointers to debug this further. 

More info:

```
uname -a 
Linux mylaptopname 5.17.2-051702-generic #202204081338 SMP PREEMPT Fri Apr 8 13:43:51 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
```


Snip from `/var/log/kern.log`
```
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043286] ================================================================================
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043291] UBSAN: array-index-out-of-bounds in /home/kernel/COD/linux/drivers/acpi/acpica/dswexec.c:401:12
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043298] index -1 is out of range for type 'acpi_operand_object *[9]'
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043302] CPU: 8 PID: 14408 Comm: kworker/u40:17 Tainted: P           OE     5.17.2-051702-generic #202204081338
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043308] Hardware name: Micro-Star International Co., Ltd. Stealth GS66 12UGS/MS-16V5, BIOS E16V5IMS.106 01/12/2022
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043311] Workqueue: events_unbound async_run_entry_fn
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043322] Call Trace:
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043325]  <TASK>
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043331]  show_stack+0x52/0x58
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043343]  dump_stack_lvl+0x4c/0x63
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043348]  dump_stack+0x10/0x12
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043351]  ubsan_epilogue+0x9/0x45
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043354]  __ubsan_handle_out_of_bounds.cold+0x44/0x49
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043359]  acpi_ds_exec_end_op+0x1a1/0x798
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043366]  acpi_ps_parse_loop+0x587/0x660
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043372]  acpi_ps_parse_aml+0x1af/0x552
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043376]  acpi_ps_execute_method+0x208/0x2ca
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043380]  acpi_ns_evaluate+0x34e/0x4f0
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043387]  acpi_evaluate_object+0x18e/0x3b4
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043390]  ? kfree+0xab/0x410
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043397]  __acpi_power_off+0x26/0x70
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043404]  acpi_power_off_unlocked+0x50/0x90
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043409]  acpi_power_off_list.isra.0+0x43/0x90
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043415]  acpi_power_transition+0xa9/0xf0
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043421]  acpi_device_set_power+0x2cb/0x360
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043428]  acpi_pci_set_power_state+0x98/0x130
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043435]  pci_set_power_state+0x69/0x1a0
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043441]  pci_prepare_to_sleep+0x6f/0xd0
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043446]  ? pci_pm_suspend_late+0x40/0x40
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043450]  pci_pm_suspend_noirq+0xda/0x2e0
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043453]  ? pci_pm_suspend_late+0x40/0x40
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043456]  dpm_run_callback+0x69/0x160
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043461]  __device_suspend_noirq+0x8f/0x280
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043465]  async_suspend_noirq+0x23/0x60
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043468]  async_run_entry_fn+0x30/0x120
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043472]  process_one_work+0x21c/0x3f0
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043479]  worker_thread+0x50/0x3d0
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043485]  ? rescuer_thread+0x390/0x390
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043491]  kthread+0xeb/0x120
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043496]  ? kthread_complete_and_exit+0x20/0x20
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043501]  ret_from_fork+0x1f/0x30
Apr 17 16:52:59 mylaptopname kernel: [ 2478.043510]  </TASK>
```

**Comment by leodotcloud** [comment]
Looks like there is a fix already: https://github.com/acpica/acpica/issues/725?

**Comment by daniel-ziegler** [comment]
Do you mean #745?

**Comment by leodotcloud** [comment]
yes, my bad, typo.

**Comment by satmandu** [comment]
This adaptation of #745 fixes this for me with kernel 5.18.1:

```
--- a/drivers/acpi/acpica/dswexec.c	2022-06-03 09:37:48.014257228 -0400
+++ b/drivers/acpi/acpica/dswexec.c	2022-06-03 09:33:58.393261022 -0400
@@ -389,9 +389,11 @@ acpi_status acpi_ds_exec_end_op(struct a
 
 		/*
 		 * All opcodes require operand resolution, with the only exceptions
-		 * being the object_type and size_of operators.
+		 * being the object_type and size_of operators as well as operands
+		 * that take no arguments.
 		 */
-		if (!(walk_state->op_info->flags & AML_NO_OPERAND_RESOLVE)) {
+		if (!(walk_state->op_info->flags & AML_NO_OPERAND_RESOLVE) &&
+				(walk_state->op_info->flags & AML_HAS_ARGS)) {
 
 			/* Resolve all operands */
```

**Comment by the-walking-code** [comment]
I constantly run into this issue with my laptop freezing as well. 

@satmandu, could you please explain how to apply the fix above so we don't have to wait until it would be pushed upstream? 

I do have some software background, but not in the linux world (I'm running Pop!_OS 22.04 LTS, one of Ubuntu derivatives).

I was trying to search for file `dswexec.c` in the system, but couldn't find it. The only reference I found is in 
`/usr/src/linux-headers-5.17.15-76051715/drivers/acpi/acpica/Makefile:24:	dswexec.o	\` which looks like this:

```
# SPDX-License-Identifier: GPL-2.0
#
# Makefile for ACPICA Core interpreter
#

ccflags-y			:= -Os -D_LINUX -DBUILDING_ACPICA
ccflags-$(CONFIG_ACPI_DEBUG)	+= -DACPI_DEBUG_OUTPUT

# use acpi.o to put all files here into acpi.o modparam namespace
obj-y	+= acpi.o

acpi-y :=		\
	dsargs.o	\
	dscontrol.o	\
	dsdebug.o	\
	dsfield.o	\
	dsinit.o	\
	dsmethod.o	\
	dsmthdat.o	\
	dsobject.o	\
	dsopcode.o	\
	dspkginit.o	\
	dsutils.o	\
	dswexec.o	\
	dswload.o	\
	dswload2.o	\
	dswscope.o	\
	dswstate.o

acpi-y +=		\
	evevent.o	\
......etc
```


My guess is that we should clone acpi repo, make a change, compile it and then somehow flash it? Again, sorry if I'm saying complete nonsense. Appreciate any help in advance.

**Comment by qzed** [comment]
@the-walking-code ACPICA is included as part of the kernel. You'd have to patch and compile the kernel yourself.

**Comment by satmandu** [comment]
@the-walking-code Are you familiar with building kernels?

e.g.
```
curl -OLf https://git.kernel.org/torvalds/t/linux-5.19-rc4.tar.gz
curl -Ls https://github.com/acpica/acpica/files/9031293/dswexec.diff.txt -o dswexec.diff
tar xf linux-5.19-rc4.tar.gz
cd linux-5.19-rc4
patch -p1 -i ../dswexec.diff
```

Now do your normal kernel build process.

[dswexec.diff.txt](https://github.com/acpica/acpica/files/9031293/dswexec.diff.txt)

**Comment by satmandu** [comment]
This DOES solve my issue though. Should I open a kernel bug and offer this as a proposed patch authored by 
@amadeuszslawinski-intel?

**Comment by pma1** [comment]
The pull #745 does fix the call trace, but not the s2idle issue. The s2idle issue might be related to i915 driver.

**Comment by sclarkson** [comment]
#745 also fixes the problem for me. Please get this into mainline linux, and the stable branches.

**Comment by RafalSkolasinski** [comment]
Is there any reason why #745 is hanging there not being merged?

**Comment by satmandu** [comment]
I'm also trying to get that upstreamed here:

https://bugzilla.kernel.org/show_bug.cgi?id=216266

**Comment by RafalSkolasinski** [comment]
Could it be that because it only affects small number of users it's considered a very low priority?

**Comment by satmandu** [comment]
Maybe? Feel free to leave a comment and offer input if you've tested this patch?

**Comment by acpibob** [comment]
Well, that is part of it.  Rafael was going to take a look at the patch, and apparently has not gotten around to it.

**Comment by maciejwieczorretman** [comment]
Looks like #745 fixed the issue, closing
